streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionStoreReceiveProcessorSupplier.java

```diff
@@ -0,0 +1,112 @@
+import org.apache.kafka.streams.state.TimestampedKeyValueStore;
+import org.apache.kafka.streams.state.ValueAndTimestamp;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class SubscriptionStoreReceiveProcessorSupplier<K, KO>
+    implements ProcessorSupplier<KO, SubscriptionWrapper<K>> {
+    private static final Logger LOG = LoggerFactory.getLogger(SubscriptionStoreReceiveProcessorSupplier.class);
+
+    private final StoreBuilder<TimestampedKeyValueStore<Bytes, SubscriptionWrapper<K>>> storeBuilder;
+    private final CombinedKeySchema<KO, K> keySchema;
+
+    public SubscriptionStoreReceiveProcessorSupplier(
+        final StoreBuilder<TimestampedKeyValueStore<Bytes, SubscriptionWrapper<K>>> storeBuilder,
+        final CombinedKeySchema<KO, K> keySchema) {
+
+        this.storeBuilder = storeBuilder;
+        this.keySchema = keySchema;
+    }
+
+    @Override
+    public Processor<KO, SubscriptionWrapper<K>> get() {
+
+        return new AbstractProcessor<KO, SubscriptionWrapper<K>>() {
+
+            private TimestampedKeyValueStore<Bytes, SubscriptionWrapper<K>> store;
+            private StreamsMetricsImpl metrics;
+            private Sensor skippedRecordsSensor;
+
+            @Override
+            public void init(final ProcessorContext context) {
+                super.init(context);
+                final InternalProcessorContext internalProcessorContext = (InternalProcessorContext) context;
+
+                metrics = internalProcessorContext.metrics();
+                skippedRecordsSensor = ThreadMetrics.skipRecordSensor(metrics);
+                store = internalProcessorContext.getStateStore(storeBuilder);
+            }
+
+            @Override
+            public void process(final KO key, final SubscriptionWrapper<K> value) {
+                if (key == null) {
+                    LOG.warn(
+                        "Skipping record due to null foreign key. value=[{}] topic=[{}] partition=[{}] offset=[{}]",
+                        value, context().topic(), context().partition(), context().offset()
+                    );
+                    skippedRecordsSensor.record();
+                    return;
+                }
+                if (value.getVersion() != SubscriptionWrapper.CURRENT_VERSION) {
+                    //Guard against modifications to SubscriptionWrapper. Need to ensure that there is compatibility
+                    //with previous versions to enable rolling upgrades. Must develop a strategy for upgrading
+                    //from older SubscriptionWrapper versions to newer versions.
+                    throw new UnsupportedVersionException("SubscriptionWrapper is of an incompatible version.");
+                }
+
+                final Bytes subscriptionKey = keySchema.toBytes(key, value.getPrimaryKey());
+
+                final ValueAndTimestamp<SubscriptionWrapper<K>> newValue = ValueAndTimestamp.make(value, context().timestamp());
```

Comment:
I didn't do anything with this, but it occurred to me that we might be able to do some optimizations if the subscription hasn't changed... WDYT?